data_safe_haven/infrastructure/programs/sre/dns_sidecar.py

```diff
@@ -0,0 +1,304 @@
+from typing import Protocol
+
+from pulumi import ComponentResource, Input, Output, ResourceOptions
+from pulumi_azure_native import authorization, containerinstance, storage
+from pulumi_azure_native.app.v20250101 import (
+    AccessMode,
+    AppLogsConfigurationArgs,
+    AzureFilePropertiesArgs,
+    ContainerArgs,
+    ContainerResourcesArgs,
+    EnvironmentVarArgs,
+    Job,
+    JobConfigurationArgs,
+    JobConfigurationScheduleTriggerConfigArgs,
+    JobTemplateArgs,
+    LogAnalyticsConfigurationArgs,
+    ManagedEnvironment,
+    ManagedEnvironmentsStorage,
+    ManagedEnvironmentStoragePropertiesArgs,
+    ManagedServiceIdentityArgs,
+    ManagedServiceIdentityType,
+    StorageType,
+    TriggerType,
+    VnetConfigurationArgs,
+    VolumeArgs,
+    VolumeMountArgs,
+    WorkloadProfileArgs,
+)
+from pulumi_azure_native.managedidentity import UserAssignedIdentity
+
+from data_safe_haven.functions import seeded_uuid
+from data_safe_haven.infrastructure.components import (
+    FileShareFile,
+    FileShareFileProps,
+    LocalDnsRecordComponent,
+    WrappedLogAnalyticsWorkspace,
+)
+from data_safe_haven.resources import resources_path
+from data_safe_haven.utility import FileReader
+
+
+class SupportsDnsMonitoring(Protocol):
+    dns_record_name: str
+    container_group_name: str
+    local_dns: LocalDnsRecordComponent
+    container_group: containerinstance.ContainerGroup
+
+
+class DnsSidecarProps:
+    """Properties of the DnsMonitorProps"""
+
+    def __init__(
+        self,
+        container_instances: list[SupportsDnsMonitoring],
+        infrastructure_subnet_id: Input[str],
+        location: Input[str],
+        log_analytics_workspace: Input[WrappedLogAnalyticsWorkspace],
+        resource_group_name: Input[str],
+        sre_fqdn: Input[str],
+        subscription_id: Input[str],
+        storage_account_name: Input[str],
+        storage_account_key: Input[str],
+    ):
+        self.container_instances = container_instances
+        self.infrastructure_subnet_id = infrastructure_subnet_id
+        self.location = location
+        self.log_analytics_workspace = log_analytics_workspace
+        self.resource_group_name = resource_group_name
+        self.sre_fqdn = sre_fqdn
+        self.subscription_id = subscription_id
+        self.storage_account_name = storage_account_name
+        self.storage_account_key = storage_account_key
+
+
+class DnsSidecarComponent(ComponentResource):
+
+    def __init__(
+        self,
+        name: str,
+        stack_name: str,
+        props: DnsSidecarProps,
+        opts: ResourceOptions | None = None,
+    ):
+        super().__init__("dsh:sre:DnsMonitorComponent", name, {}, opts)
+        child_opts = ResourceOptions.merge(opts, ResourceOptions(parent=self))
+
+        file_share = storage.FileShare(
+            f"{self._name}_file_share_dnsmonitor",
```

Comment:
```suggestion
            f"{self._name}_file_share_dnssidecar",
```
We should be consistent about what we call this new component. Either `monitor` or `sidecar` is fine with me.

Reply:
Let's stay with DNS Sidecar. I will change the code accordingly